parfive/downloader.py

```diff
@@ -0,0 +1,436 @@
+
+        # Setup queues
+        self.http_queue = asyncio.Queue(loop=self.loop)
+        self.http_tokens = asyncio.Queue(maxsize=self.max_conn, loop=self.loop)
+        self.ftp_queue = asyncio.Queue(loop=self.loop)
+        self.ftp_tokens = asyncio.Queue(maxsize=self.max_conn, loop=self.loop)
+        for i in range(self.max_conn):
+            self.http_tokens.put_nowait(Token(i + 1))
+            self.ftp_tokens.put_nowait(Token(i + 1))
+
+    def enqueue_file(self, url, path=None, filename=None, overwrite=None, **kwargs):
+        """
+        Add a file to the download queue.
+
+        Parameters
+        ----------
+
+        url : `str`
+            The URL to retrieve.
+
+        path : `str`
+            The directory to retrieve the file into.
+
+        filename : `str` or `callable`
+            The filename to save the file as. Can also be a callable which
+            takes two arguments the url and the response object from opening
+            that URL, and returns the filename. (Note, for FTP downloads the
+            response will be ``None``.)
+
+        chunksize : `int`
+            The size (in bytes) of the chunks to be downloaded for HTTP downloads.
+
+        overwrite : `bool` or `str`, optional
+            Determine how to handle downloading if a file already exists with the
+            same name. If `False` the file download will be skipped and the path
+            returned to the existing file, if `True` the file will be downloaded
+            and the existing file will be overwritten, if `'unique'` the filename
+            will be modified to be unique.
+
+        kwargs : `dict`
+            Extra keyword arguments are passed to `aiohttp.ClientSession.get`
+            or `aioftp.ClientSession` depending on the protocol.
+        """
+        overwrite = overwrite or self.overwrite
+
+        if path is None and filename is None:
+            raise ValueError("either path or filename must be specified.")
+        if not filename:
+            filepath = partial(default_name, path)
+        elif callable(filename):
+            filepath = filename
+        else:
+            # Define a function because get_file expects a callback
+            def filepath(*args):
+                return filename
+
+        if url.startswith("http"):
+            get_file = partial(self._get_http, url=url, filepath_partial=filepath,
+                               overwrite=overwrite, **kwargs)
+            self.http_queue.put_nowait(get_file)
+        elif url.startswith("ftp"):
+            if aioftp is None:
+                raise ValueError("The aioftp package must be installed to download over FTP")
+            get_file = partial(self._get_ftp, url=url, filepath_partial=filepath,
+                               overwrite=overwrite, **kwargs)
+            self.ftp_queue.put_nowait(get_file)
+        else:
+            raise ValueError("url must start with either http or ftp")
```

Comment:
```suggestion
            raise ValueError("URL must start with HTTP or FTP.")
```
maybe?

Reply:
No, the check is actually case sensitive. I am not sure what the best way to do that is?

Reply:
Oh, in that case, not a clue. Just ignore my comment.

Reply:
> No, the check is actually case sensitive. I am not sure what the best way to do that is?

I think it's best to use `urlparse` here.

```python
>>> from urllib.parse import urlparse
>>> urlparse('http://google.com').scheme
'http'
>>> urlparse('HTTP://google.com').scheme
'http'
>>> urlparse('FTP://google.com').scheme
'ftp'
```

Reply:
Also, what about https? Why would this not work with https downloads?

Reply:
`"https".startswith("http") == True` ;)